ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -0,0 +1,350 @@
+   * @param chunkBuffer The PutChunk buffer to compress.
+   * @param isFullChunk Whether this is a full size chunk (4MB) or smaller.
+   * @return Returns a new compressed buffer.  It returns null is no compression is applied.
+   */
+  public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
+    Objects.requireNonNull(chunkBuffer, "blobProperties cannot be null.");
+
+    // Check the blob size.  If it's too small, do not compress.
+    int sourceDataSize = chunkBuffer.readableBytes();
+    if (sourceDataSize < compressionConfig.minimalSourceDataSizeInBytes) {
+      logger.info("No compression applied because source data size " + sourceDataSize + " is smaller than minimal size "
+          + compressionConfig.minimalSourceDataSizeInBytes);
+      // Emit metrics to count how many compressions skipped due to small size.
+      compressionMetrics.compressSkipRate.mark();
+      compressionMetrics.compressSkipSizeTooSmall.inc();
+      return null;
+    }
+
+    // Convert ByteBuf to byte[] before applying compression.
+    byte[] sourceBuffer;
+    int sourceBufferOffset;
+    try {
+      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
+      sourceBuffer = bufferInfo.getLeft();
+      sourceBufferOffset = bufferInfo.getMiddle();
+      sourceDataSize = bufferInfo.getRight();
+    } catch (Exception ex) {
+      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
+              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
+          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
+
+      // Emit metrics for count how many compressions skipped due to buffer conversion error.
+      compressionMetrics.compressErrorRate.mark();
+      compressionMetrics.compressErrorBufferConversion.inc();
+      return null;
+    }
+
+    // Get default compressor.  If config missing, fallback to Zstandard.
+    String algorithmName = compressionConfig.algorithmName;
+    Compression defaultCompressor = allCompressions.get(algorithmName);
+    if (defaultCompressor == null) {
+      // Emit metrics for default compressor name does not exist.
+      compressionMetrics.compressErrorConfigInvalidCompressorName.inc();
+      logger.error("The default compressor algorithm, " + algorithmName
+          + ", specified in config does not exist.  This config is ignored.");
+      algorithmName = ZstdCompression.ALGORITHM_NAME;
+      defaultCompressor = allCompressions.get(algorithmName);
+    }
+
+    // Apply compression.
+    CompressionMetrics.AlgorithmMetrics algorithmMetrics = compressionMetrics.getAlgorithmMetrics(algorithmName);
+    Pair<Integer, byte[]> compressResult;
+    try {
+      algorithmMetrics.compressRate.mark();
+      long startTime = System.nanoTime();
+      compressResult = defaultCompressor.compress(sourceBuffer, sourceBufferOffset, sourceDataSize);
```

Comment:
at same point we have to create ByteBuf outside of compress method. Do you have a plan for that?

Reply:
I need more info about this comment.  Currently compression service does not create ByteBuf, the caller (PutOperation and GetOperation) will capture the method output, byte[], and convert it to ByteBuf or ByteBuffer, whatever works for them.